noir-projects/noir-contracts/contracts/escrowable_token_contract/src/types/balances_map.nr

```diff
@@ -58,59 +60,66 @@ impl<T> BalancesMap<T, ()> {
 }
 
 impl<T> BalancesMap<T, &mut PrivateContext> {
+    /**
+     * Adds a note to the balance of the owner.
+     *
+     * @param owner: The address of the note owner.
+     * @param addend: The amount to add to the balance
+     * @param incoming_viewer_and_nullifier: The address to use for `ivpk` and `npk` lookups
+     * @param outgoing_viewer: The address to use for `ovpk` lookup
+     */
     pub fn add<T_SERIALIZED_LEN, T_SERIALIZED_BYTES_LEN>(
         self: Self,
-        outgoing_viewer: AztecAddress,
         owner: AztecAddress,
-        addend: U128
+        addend: U128,
+        incoming_viewer_and_nullifier: AztecAddress,
+        outgoing_viewer: AztecAddress
     ) where T: NoteInterface<T_SERIALIZED_LEN, T_SERIALIZED_BYTES_LEN> + OwnedNote {
         let context = self.map.context;
         let header = context.get_header();
-        let outgoing_viewer_ovpk_m = header.get_ovpk_m(context, outgoing_viewer);
-        let owner_ivpk_m = header.get_ivpk_m(context, owner);
 
-        // We fetch the nullifier public key hash from the registry / from our PXE
-        let owner_npk_m_hash = header.get_npk_m_hash(context, owner);
-        let mut addend_note = T::new(addend, owner_npk_m_hash);
+        let ivpk_m = header.get_ivpk_m(context, incoming_viewer_and_nullifier);
+        let npk_m_hash = header.get_npk_m_hash(context, incoming_viewer_and_nullifier);
+        let ovpk_m = header.get_ovpk_m(context, outgoing_viewer);
+
+        let mut addend_note = T::new(addend, npk_m_hash);
 
-        // docs:start:insert
-        self.map.at(owner).insert(&mut addend_note, outgoing_viewer_ovpk_m, owner_ivpk_m);
-        // docs:end:insert
+        self.map.at(owner).insert(&mut addend_note, ovpk_m, ivpk_m);
     }
 
+    /**
+     * Removes notes from the balance of the owner until removing at least `subtrahend`
+     * If removing "too much" adds a note to the balance of the owner with the diff.
+     *
+     * @param owner: The address of the note owner.
+     * @param subtrahend: The amount to remove from the balance
+     * @param incoming_viewer_and_nullifier: The address to use for `ivpk` and `npk` lookups
+     * @param outgoing_viewer: The address to use for `ovpk` lookup
+     */
     pub fn sub<T_SERIALIZED_LEN, T_SERIALIZED_BYTES_LEN>(
         self: Self,
         owner: AztecAddress,
-        subtrahend: U128
+        subtrahend: U128,
+        incoming_viewer_and_nullifier: AztecAddress,
+        outgoing_viewer: AztecAddress
     ) where T: NoteInterface<T_SERIALIZED_LEN, T_SERIALIZED_BYTES_LEN> + OwnedNote {
-        // docs:start:get_notes
         let options = NoteGetterOptions::with_filter(filter_notes_min_sum, subtrahend);
         let maybe_notes = self.map.at(owner).get_notes(options);
```

Comment:
ref my note in main.nr about how i get get_notes should be on the `outgoing_viewer` instead.